ДЗ по "Основы функциональной разработки"

v-nvtsk opened the pull request.

v-nvtsk committed:
build: add config

v-nvtsk committed:
feat: add curry

v-nvtsk committed:
feat: add semver-sort

v-nvtsk committed:
feat: add spiral

v-nvtsk committed:
feat: add parallel

v-nvtsk committed:
feat: add sum

v-nvtsk committed:
refactor: add tests for Parallel

v-nvtsk committed:
style: clear comments

v-nvtsk committed:
feat: add examples to index.ts

v-nvtsk committed:
refactor: parallel

remove repeating code
rename functions to more readable

v-nvtsk committed:
fix: add missed sum .toString test

v-nvtsk committed:
refactor: parallel - add catch for rejected results

vvscode commented on .husky/pre-commit:
```diff
@@ -0,0 +1,4 @@
+#!/usr/bin/env sh
+. "$(dirname -- "$0")/_/husky.sh"
+
+npm test
```
тесты на прекоммит не очень идея (потому что медленно и пропадает желание делать коммиты часто), может быть на pre-push (но там нужен lint-* другой)

luckfinder05 replied:
Действительно очень неудобно, особенно, когда покрытие не соответствует требуемому - приходится коммитить с --no-verify или коммитить уже по завершению задания.
Нашёл lint-prepush, но последний раз он обновлялся 2 года назад.
Добавил, заработал только когда жёстко ограничил проверку tsc файлами ./src/**/*.ts Если проверять все ts файлы - заваливал проверку на конфигурационных файлах.

vvscode commented on .eslintrc.js:
```diff
@@ -0,0 +1,36 @@
+      "error",
+      {
+        code: 120,
+        comments: 80,
```
странная настройка - код может быть шире комментариев?

luckfinder05 replied:
Настраивал по мотивам интернетных гайдов.
сам недоумеваю - очень неудобно.
Поменял на 120.

vvscode commented on src/curry.test.ts:
```diff
@@ -0,0 +1,40 @@
+    );
+  });
+
+  let hof: any;
```
хорошо бы тут разобраться с типизацией

luckfinder05 replied:
Описал тип Hof

vvscode commented on src/curry.ts:
```diff
@@ -0,0 +1,13 @@
+// Написать функцию для каррирования (https://ru.wikipedia.org/wiki/%D0%9A%D0%B0%D1%80%D1%80%D0%B8%D1%80%D0%BE%D0%B2%D0%B0%D0%BD%D0%B8%D0%B5)
+// Пример использования функции
+
+type Fn = (...args: any[]) => any;
```
а чем это от `Function` отличается?

luckfinder05 replied:
eslint ругается на Function:
Don't use `Function` as a type. The `Function` type accepts any function-like value.
It provides no type safety when calling the function, which can be a common source of bugs.
It also accepts things like class declarations, which will throw at runtime as they will not be called with `new`.
If you are expecting the function to accept certain arguments, you should explicitly define the function shape.eslint[@typescript-eslint/ban-types](https://typescript-eslint.io/rules/ban-types)

vvscode replied:
ну так ты почитай описание `It provides no type safety ` -  а чем твое лучше?

vvscode commented on src/parallel.test.ts:
```diff
@@ -0,0 +1,116 @@
+  });
+
+  it("should return result ignoring rejected promises", async () => {
+    const result = await new Parallel(1).jobs(
```
я вижу тесты на результат выполнение, но я не вижу тестов на то, что число потоков и правда ограничено - как так?

luckfinder05 replied:
Сделал число активных потоков публичным свойством и протестировал.
Их число физически ограничено при инициализации - в методе jobs в цикле создаётся заданное количество потоков, при завершении каждого из них запускается следующий из списка.
Но, если бы оставил как есть - нельзя было бы запускать повторно или добавлять задачи в процессе обработки.
Сейчас исправил.

vvscode replied:
я все еще не вижу тестов на число потоков - мне тебе верить на слово? =)
про свойство - ты проверил, что свойство установилось, это здорово, но это не значит, что лимит действительно такой

очень советую разобраться как простетировать это корректно (именно с числом задач)

vvscode commented on src/parallel.ts:
```diff
@@ -0,0 +1,55 @@
+      return;
+    }
+    const task = taskList.shift();
+    new Promise((resolve) => {
```
а в чем смысл new Promise? что мешаело сделать `return task().then...`

luckfinder05 replied:
Убрал. То, зачем его добавлял можно вполне выполнить в .finally

vvscode commented on src/parallel.ts:
```diff
@@ -0,0 +1,55 @@
+
+  private async getFinalResult(): Promise<[]> {
+    return new Promise((resolve) => {
+      this.onReadycb = resolve.bind(this.result);
```
а если я захочу подписаться после того как джобы отработали? и тест на этот случай, пожалуйста

luckfinder05 replied:
Не понял на что подписаться после того, как джобы отработали?
Это внутренний метод, возвращающий результат выполнения промиса .jobs().
Возможно связано с вопросом https://github.com/v-nvtsk/otus-jsbasic-dz27-fp/pull/1/files#r1468823037

vvscode replied:
да, это я не досмотрел

тогда давай чинить типизацию, сейчас это просто пустой массив, а это же не так

vvscode commented on src/sort.test.ts:
```diff
@@ -0,0 +1,36 @@
+import { semverSort } from "./sort";
+
+global.console = {
```
jest.spyOn

luckfinder05 replied:
ок. спасибо.

vvscode commented on src/sort.ts:
```diff
@@ -0,0 +1,19 @@
+// Реализовать функцию, реализующую сортировку с учетом правил semver
+
+function checkDigit(num: number | undefined): number {
```
Number.isInteger(1) ?

luckfinder05 replied:
Здесь она мне не поможет, она возвращает boolean.
Может быть имеет смысл назвать мою функцию иначе, но она не только проверяет, но и возвращает число или ноль.

vvscode replied:
функцию точно имеет смысл назвать иначе =)

vvscode commented on src/sort.ts:
```diff
@@ -0,0 +1,19 @@
+
+export function semverSort(arr: string[]): any[] {
+  return arr
+    .map((el: string): number[] => el.split(".").map(Number))
```
а если точек больше 2 ?

luckfinder05 replied:
Имеется ввиду   '1.2.3.4.5.6.7'?
Работает. Хотя семвер подразумевает только 3 цифры вроде?

vvscode commented on src/sum.ts:
```diff
@@ -0,0 +1,22 @@
+  };
+
+  add.valueOf = () => {
+    return args.reduce((acc: number, curr: number) => acc + curr, 0);
```
а если не пладить массивы? а чуть поменьше памяти расходовать - сможем?

luckfinder05 replied:
Это не смог придумать.

profAlexGit replied:
Давай отталкиваться от того, что в примере sum вызывается либо без аргументов, либо с одним.
Последующие вызовы происходят с одним аргументом.
При каждом вызове, мы суммируем новое значение с предыдущим результатом и снова возвращаем функцию. При этом переопределяем toString и valueOf (не составит труда, т.к. значение у нас есть).

v-nvtsk committed:
refactor: add jest spyOn

v-nvtsk committed:
fix: types in curry

v-nvtsk committed:
build: change eslint allowed comments width

v-nvtsk committed:
fix: parallel

add functions getTaskCount, getActiveJobsCount

remove needless promise in runNewJob

add ability to add jobs after finishing

add ability to add jobs while processing

add tests to control actviveJobs count

v-nvtsk committed:
style: fix files formatting

v-nvtsk committed:
build: add lint-prepush

v-nvtsk committed:
fix: semver for infinite nested numeration

v-nvtsk committed:
fix: prevent run Parallel jobs without jobs

v-nvtsk committed:
fix: add parallel tests

add test for running jobs without jobs

add tests for count active jobs in the moment

add test for console.error on Error in task

v-nvtsk committed:
fix test for parallel exactly 2 jobs

v-nvtsk committed:
fix curry function return type

v-nvtsk committed:
fix: sum rid of arrays

v-nvtsk committed:
fix: index.ts minor changes for sum sample

v-nvtsk committed:
fix curry function return type

v-nvtsk committed:
fix: sort - rename checkDigit to getValidNumber

v-nvtsk committed:
add github workflow
